### PR TITLE
KubernetesExecutor multi_namespace_mode can use namespace list to avoid requiring cluster role

### PR DESCRIPTION
**Pull request by XD-DENG**
Currently `KubernetesExecutor`'s `multi_namespace_mode` requires the Scheduler to have cluster-scope role on the Kubernetes Cluster, because it's using function `list_pod_for_all_namespaces()`.

However, in certain enterprise environments, it's not possible for users to have cluster-scope role. For example, they may only get permissions in a namespace, rather on the whole cluster. Always allowing the Scheduler pod to have cluster-scope role is not a good from security aspect either.

This change aims to make `KubernetesExecutor`'s `multi_namespace_mode` work without cluster-scope role.

(This was discussed at the mail list at https://lists.apache.org/thread/xxsppw7qwvky78l6nx41vlz593gj4zqb)

I'm sure folks would have suggestions and we need to future refine this change, but I would like to bring up the discussion by creating this PR first.

**UPDATE:**
Advantages this change brings:
- Better fits enterprise environment
- Better security: limit the permissions that the Scheduler Pod needs, so that it doesn't have too much permissions which it doesn't have to have (earlier it has to have a cluster role in order to use `multi_namespace_mode`)

### CODE REVIEW

**Comment by XD-DENG** [comment]
Hi @potiuk and @ferruzzi tagging you both, to follow up [our earlier discussion in the mail list](https://lists.apache.org/thread/xxsppw7qwvky78l6nx41vlz593gj4zqb).

File: airflow/executors/kubernetes_executor.py [hunk 1]
```diff
@@ -66,7 +67,7 @@ class ResourceVersion:
     """Singleton for tracking resourceVersion from Kubernetes"""
 
     _instance = None
-    resource_version = "0"
+    resource_version: dict[str | None, str | None] = {}
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py]: does this not work?
```suggestion
    resource_version: dict[str, str] = {}
```

File: airflow/executors/kubernetes_executor.py [hunk 2]
```diff
@@ -288,15 +294,33 @@ def _make_kube_watcher(self) -> KubernetesJobWatcher:
         watcher.start()
         return watcher
 
-    def _health_check_kube_watcher(self):
-        if self.kube_watcher.is_alive():
-            self.log.debug("KubeJobWatcher alive, continuing")
-        else:
-            self.log.error(
-                "Error while health checking kube watcher process. Process died for unknown reasons"
+    def _make_kube_watchers(self) -> dict[str | None, KubernetesJobWatcher]:
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py]:
```suggestion
    def _make_kube_watchers(self) -> dict[str, KubernetesJobWatcher]:
```

**Review by ferruzzi — APPROVED** [review]
Other than the comments already made, LGTM

File: airflow/config_templates/default_airflow.cfg [hunk 3]
```diff
@@ -1197,9 +1197,14 @@ delete_worker_pods_on_failure = False
...
 
 # Allows users to launch pods in multiple namespaces.
-# Will require creating a cluster-role for the scheduler
+# Will require creating a cluster-role for the scheduler, or use namespace_list configuration.
 multi_namespace_mode = False
 
+# If multi_namespace_mode is True while scheduler does not have a cluster-role,
+# give the list of namespaces where the scheduler will schedule jobs
+# Scheduler needs to have the necessary permissions in these namespaces.
+namespace_list =
```
Comment (dstandish) [inline airflow/config_templates/default_airflow.cfg]: maybe this would be better.  

```suggestion
multi_namespace_mode_namespace_list =
```
it's really a sub-option of multi_namespace_mode.

wdyt?

File: airflow/executors/kubernetes_executor.py (hunk 1 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: Hi @dstandish , when `multi_namespace_mode` is `True` & `namespace_list` is not specified, that means the user will still want to apply the way using cluster-scope role. In this case, argument `namespace` will be None for the corresponding `KubernetesJobWatcher`. That's they the key can be `None`.

But the value doesn't have to be `str | None`. So I will change this to `resource_version: dict[str | None, str] = {}`.

File: airflow/executors/kubernetes_executor.py (hunk 2 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: The same as clarified above.

File: airflow/config_templates/default_airflow.cfg (hunk 3 above)
Comment (XD-DENG) [reply airflow/config_templates/default_airflow.cfg]: Sure, that sounds sensible to me.

Let me add a follow-up commit.

Comment (XD-DENG) [reply airflow/config_templates/default_airflow.cfg]: Address with commit https://github.com/apache/airflow/pull/28047/commits/1c1f25d92e0faeb4f84069056dbdc24c1e53fa93

File: airflow/executors/kubernetes_executor.py (hunk 1 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: Changed to `resource_version: dict[str | None, str] = {}` in commit
https://github.com/apache/airflow/pull/28047/commits/a23c33e90bd6666286e03cc676aba9a6322d570c

**Comment by XD-DENG** [comment]
Hi @dstandish , clarified/addressed your earlier comments. Pls help take another look when you get time? Thanks a lot!

File: airflow/executors/kubernetes_executor.py [hunk 4]
```diff
@@ -140,9 +141,14 @@ def _run(
...
-                watcher.stream, kube_client.list_pod_for_all_namespaces, **kwargs
-            )
+            if self.kube_config.multi_namespace_mode_namespace_list:
+                list_worker_pods = functools.partial(
+                    watcher.stream, kube_client.list_namespaced_pod, self.namespace, **kwargs
+                )
+            else:
+                list_worker_pods = functools.partial(
+                    watcher.stream, kube_client.list_pod_for_all_namespaces, **kwargs
+                )
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py]: might be cleaner to move this if block to a private method and i think it can be simplified to this:
```python
    def _list_worker_pods(self, watcher: watch.Watch, kube_client, **kwargs):
        if self.namespace is None:
            return watcher.stream(kube_client.list_pod_for_all_namespaces, **kwargs)
        else:
            return watcher.stream(kube_client.list_namespaced_pod, self.namespace, **kwargs)
```

then used like this
```python
        for event in self._list_worker_pods(watcher, kube_client, **kwargs):
```

I think this makes sense because, this class doesn't really need to know about the airflow config.  It's just a watcher.  And, it's either watching a namespace, or watching all namespaces.  If namespace is none, it watches all; otherwise, it watches the namespace.

File: airflow/kubernetes/kube_config.py [hunk 5]
```diff
@@ -57,6 +57,14 @@ def __init__(self):
...
         self.kube_namespace = conf.get(self.kubernetes_section, "namespace")
         self.multi_namespace_mode = conf.getboolean(self.kubernetes_section, "multi_namespace_mode")
+        if self.multi_namespace_mode and conf.get(
+            self.kubernetes_section, "multi_namespace_mode_namespace_list"
+        ):
+            self.multi_namespace_mode_namespace_list = conf.get(
+                self.kubernetes_section, "multi_namespace_mode_namespace_list"
+            ).split(",")
+        else:
+            self.multi_namespace_mode_namespace_list = None
```
Comment (dstandish) [inline airflow/kubernetes/kube_config.py:67]:
```suggestion
        self.multi_namespace_mode_namespace_list = conf.getjson(
            self.kubernetes_section, "multi_namespace_mode_namespace_list"
        )
```
maybe this would be a good use of json. or could add a `getlist`

File: airflow/executors/kubernetes_executor.py [hunk 6]
```diff
@@ -444,6 +471,23 @@ def __init__(self):
...
+                    )
+            else:
+                pods = self.kube_client.list_pod_for_all_namespaces(**query_kwargs).items
+        else:
+            pods = self.kube_client.list_namespaced_pod(
+                namespace=self.kube_config.kube_namespace, **query_kwargs
+            ).items
+
+        return pods
+
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py:485]:
```suggestion
    @cached_property
    def _namespaces(self):
        if self.kube_config.multi_namespace_mode_namespace_list:
            return self.kube_config.multi_namespace_mode_namespace_list
        elif self.kube_config.multi_namespace_mode:
            return []
        else:
            return [self.kube_config.kube_namespace]

    def _list_pods(self, query_kwargs):
        if not self._namespaces:
            yield from self.kube_client.list_pod_for_all_namespaces(**query_kwargs).items
        for namespace in self._namespaces:
            yield from self.kube_client.list_namespaced_pod(namespace=namespace, **query_kwargs).items

```

File: airflow/executors/kubernetes_executor.py [hunk 7]
```diff
@@ -497,16 +541,16 @@ def clear_not_launched_queued_tasks(self, session=None) -> None:
 
             # Try run_id first
             kwargs["label_selector"] += ",run_id=" + pod_generator.make_safe_label_value(ti.run_id)
-            pod_list = self.kube_client.list_namespaced_pod(self.kube_config.kube_namespace, **kwargs)
-            if pod_list.items:
+            pod_list = self._list_pods(kwargs)
+            if pod_list:
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py:542]:
```suggestion
            for _ in self._list_pods(kwargs):
```
if you accept above suggestion, you'd have to do this

File: airflow/executors/kubernetes_executor.py [hunk 8]
```diff
@@ -497,16 +541,16 @@ def clear_not_launched_queued_tasks(self, session=None) -> None:
...
                 continue
             # Fallback to old style of using execution_date
             kwargs["label_selector"] = (
                 f"{base_label_selector},"
                 f"execution_date={pod_generator.datetime_to_label_safe_datestring(ti.execution_date)}"
             )
-            pod_list = self.kube_client.list_namespaced_pod(self.kube_config.kube_namespace, **kwargs)
-            if pod_list.items:
+            pod_list = self._list_pods(kwargs)
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py:549]: and same here

File: airflow/executors/kubernetes_executor.py [hunk 9]
```diff
@@ -287,15 +293,35 @@ def _make_kube_watcher(self) -> KubernetesJobWatcher:
...
+                    (
+                        "Error while health checking kube watcher process for namespace %s. "
+                        "Process died for unknown reasons"
+                    ),
+                    namespace,
+                )
+                ResourceVersion().resource_version[namespace] = "0"
+                self.kube_watchers[namespace] = self._make_kube_watcher(namespace)
 
     def run_next(self, next_job: KubernetesJobType) -> None:
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py:319]: in this method construct_pod is called with `self.namespace`. does this make sense with multi-namespace mode?  how are additional namespaces used if pods are always created with the one namespace?

File: airflow/executors/kubernetes_executor.py [hunk 10]
```diff
@@ -287,15 +293,35 @@ def _make_kube_watcher(self) -> KubernetesJobWatcher:
...
+                else [None]
             )
-            ResourceVersion().resource_version = "0"
-            self.kube_watcher = self._make_kube_watcher()
+        else:
+            namespaces_to_watch = [self.kube_config.kube_namespace]
+
+        for namespace in namespaces_to_watch:
+            watchers[namespace] = self._make_kube_watcher(namespace)
+        return watchers
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py]: this is somewhat redundant logic. maybe you could just calculate "namespaces" once (e.g. in the property i suggested adding to k8s exec) and then pass that along when creating this class.  

then you could do something like 
```
    def _make_kube_watchers(self) -> dict[str | None, KubernetesJobWatcher]:
        if not self.namespaces:
            return {None: self._make_kube_watcher(None)}
        return {x: self._make_kube_watcher(x) for x in self.namespaces}
```

i don't feel great about having `None` as a key in the dict... it's not a super clear as a signaling mechanism. (and,  of course, it's not a namespace, though sometimes it appears that way in the code).

maybe we'll think of something as we work through it.

Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: Hi @dstandish , regarding the key, I added https://github.com/apache/airflow/pull/28047/commits/928b8399cad70c16c451096cf28a5222c00db084 . The key will be `ALL_NAMESPACES` when `namespace_list` is not given & `multi_namespace_mode=True`. `"ALL_NAMESPACES"` is not legal as K8s namespace, so we don't have to worry about name collision with a real namespace.

Let me know how you find it, comparing with `None`?

Regarding your other comments, will need more time to get back to you (actually haven't got time to read in details yet. Busy with day works). Appreciate the review and the suggestions!

Comment (dstandish) [reply airflow/executors/kubernetes_executor.py]: > "ALL_NAMESPACES" is not legal as K8s namespace, so we don't have to worry about name collision with a real namespace.

Nice, that's great.  A little more descriptive, and very convenient that there's no risk of collision.  

re the other stuff, no worries...  and... sorry if they are too "code golfy" ... just... was looking at how to maybe streamline some parts of it... just suggestions

Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: They are definitely not "golfy" I believe! I do appreciate you taking time to help suggest on these details, and I find them extremely useful!

Will get back to you here later.

File: airflow/kubernetes/kube_config.py (hunk 5 above)
Comment (XD-DENG) [reply airflow/kubernetes/kube_config.py:67]: Hi @dstandish , I'm not sure if this is the correct thing to do. 

I think it's easier to allow users pass a value like `ns-a,ns-b,ns-c` to `multi_namespace_mode_namespace_list`, rather than asking them to put something like `"ns-a","ns-b","ns-c"`. So treating it as a json or list may not be making sense IMHO, and I would suggest we keep the existing way.

Let me know how you find it?

File: airflow/executors/kubernetes_executor.py (hunk 4 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: @dstandish  , Great point to skip checking the config and simplify the checks! I have added https://github.com/apache/airflow/pull/28047/commits/2aeb76939d1fdeecc04c1901d85244451de7d2fe to address this.

But IMHO I'm nos sure if we should add one more private method for it, as we don't re-use it anywhere, and passing the args around doesn't seem clean (I did abstract for a similar case in the `KubernetesExecutor` code, the `_list_pods` method, as it's being re-used multiple times)

File: airflow/kubernetes/kube_config.py (hunk 5 above)
Comment (dstandish) [reply airflow/kubernetes/kube_config.py:67]: sure

File: airflow/executors/kubernetes_executor.py (hunk 6 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py:485]: Sorry I don't see very obvious improvement/advantage from this change. May you clarify a bit more? Otherwise I believe the existing implementation is clearer IMHO

File: airflow/executors/kubernetes_executor.py (hunk 8 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py:549]: The lines being changes here seems outdated? They are no longer there in the latest version in this PR.

Regarding the change itself, please refer to my question above.

File: airflow/executors/kubernetes_executor.py (hunk 9 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py:319]: @dstandish  , that's a good question, and thanks for reminding on it.

Actually along preparing this PR, I kept having this question: did this `multi_namespace_mode` feature ever really work for anyone? I also noticed a few similar issues, and I shared them at https://lists.apache.org/thread/nqnq2gbvtx6fyycsqhgjhlz1yotq0vhc

Comment (dstandish) [reply airflow/executors/kubernetes_executor.py:319]: It looks like maybe if you don't set `namespace` in airflow cfg then you may be able to set it with executor config

i don't know more than that

are you using it now?

File: airflow/executors/kubernetes_executor.py (hunk 6 above)
Comment (dstandish) [reply airflow/executors/kubernetes_executor.py:485]: yeah, not really important sometimes i just think generators are a cleaner than managing the data structures

if you want to be really picky you could say that, in the case where you are using a namespace list using the generator will sometimes a few API calls because if, say, you find the pod in the first namespace, you don't need to try the others.

but i think mainly my motivation was just separation of the logic to determine namespaces from the logic to list the pods, because smaller functions can be easier to understand.

but feel free to ignore.

File: airflow/executors/kubernetes_executor.py (hunk 4 above)
Comment (dstandish) [reply airflow/executors/kubernetes_executor.py]: yeah, similar to below, i guess the idea behind the suggestion is that, code reuse isn't the only reason to extract logic into a method. sometimes it just makes the code easier to read... so you just see `for pod in list_pods()` rather than a bunch of logic in an if block.

again, feel free to disregard, just suggestion

side note though, i don't think there's any need to use functools.partial here, i don't think it is doing anything different from just calling stream() here, which returns a generator, and iterating that later.  so i think we can get rid of that (i know that's just how it was)

File: airflow/executors/kubernetes_executor.py (hunk 9 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py:319]: Our team are using KubernetesExecutor, and we do run it with with a multi namespace setting. But we couldn't use vanilla `multi_namespace_mode` as it is. That's exactly why I'm proposing the changes in this PR ;-)

Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py:319]: > It looks like maybe if you don't set namespace in airflow cfg then you may be able to set it with executor config

And yes, I believe this is the answer to the original question you shared above.

Comment (dstandish) [reply airflow/executors/kubernetes_executor.py:319]: cool

**Comment by dstandish** [comment]
you may want to look at file task handler.  when reading logs from k8s task, seems to assume single namespace

**Comment by XD-DENG** [comment]
> i just noticed something coincidentally... you may want to look at file task handler. when reading logs from k8s task, seems to assume single namespace

Thanks @dstandish , yes, I have already noticed that earlier. I plan to address that separately later though if by then other folks haven't taken it up. That part was totally missed when this `multi_namespace_mode` was introduced (again, questioning how people are using this feature so far)

File: airflow/executors/kubernetes_executor.py (hunk 4 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: Thanks again @dstandish . This part makes sense to me. I added commit https://github.com/apache/airflow/pull/28047/commits/9eaeede8dabc8f823d113ecb85112e43c31a2a8e to introduce the private method `_pod_events`, and get rid of `functool.partial()`. Let me know if you find it ok?

**Comment by dstandish** [comment]
> . That part was totally missed when this `multi_namespace_mode` was introduced (again, questioning how people are using this feature so far)

sounds good

File: airflow/executors/kubernetes_executor.py (hunk 6 above)
Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py:485]: For this part, I personally find the current coding is easier to comprehend, and no disadvantage in terms performance. So I personally prefer not to make change in this part.

Please let me know if you find it ok? Thanks

**Comment by XD-DENG** [comment]
Hi @dstandish , I tried to address your earlier comments (very nice inputs!), either changes have been made or I have clarified why I would prefer to do differently. Please let me know your thoughts.

(Sorry for getting back to you a bit slowly. Only have some time after work for this)

**Comment by dstandish** [comment]
there is nothing slow about it at all... lemme look

File: airflow/executors/kubernetes_executor.py (hunk 6 above)
Comment (dstandish) [reply airflow/executors/kubernetes_executor.py:485]: sure thing

File: tests/executors/test_kubernetes_executor.py [hunk 11]
```diff
@@ -1108,9 +1209,9 @@ def effect():
             except Exception as e:
                 assert e.args == ("sentinel",)
 
-            # both  resource_version should be 0 after _run raises and exception
+            # both resource_version should be 0 after _run raises and exception
```
Comment (dstandish) [inline tests/executors/test_kubernetes_executor.py]:
```suggestion
            # both resource_version should be 0 after _run raises an exception
```

File: airflow/executors/kubernetes_executor.py [hunk 12]
```diff
@@ -79,7 +81,7 @@ class KubernetesJobWatcher(multiprocessing.Process, LoggingMixin):
 
     def __init__(
         self,
-        namespace: str | None,
+        namespace: str,
         multi_namespace_mode: bool,
```
Comment (dstandish) [inline airflow/executors/kubernetes_executor.py]: i think this can be removed now, along with its attr

Comment (XD-DENG) [reply airflow/executors/kubernetes_executor.py]: Nice catch! Addressed with https://github.com/apache/airflow/pull/28047/commits/742468ef37a0725d657ac4790f9c3daf22df9766

**Review by dstandish — APPROVED** [review]
looks good to me. maybe worth getting second set of eyes e.g. from @jedcunningham or @uranusjr

File: airflow/executors/kubernetes_executor.py [hunk 13]
```diff
@@ -66,7 +68,7 @@ class ResourceVersion:
     """Singleton for tracking resourceVersion from Kubernetes."""
 
     _instance = None
-    resource_version = "0"
+    resource_version: dict[str, str] = {}
```
Comment (uranusjr) [inline airflow/executors/kubernetes_executor.py:71]: We can probably improve this class further, but that can be a separate PR.

**Review by uranusjr — APPROVED** [review]
I didn’t think everything _entirely_ through but this should be good.

**Comment by XD-DENG** [comment]
Thanks both @dstandish  @uranusjr ! The whole PR has improved a lot through the review.

May I also get some inputs from you about the failing test? The log isn't helping much.
Is there any known issue of the test pipeline, or just something missed/wrong here? Thanks a lot

**Comment by dstandish** [comment]
@XD-DENG @potiuk  this gives an idea.  maybe we don't need multiple watchers.  maybe a single watcher can watch multiple namespaces.

**Comment by dstandish** [comment]
and... maybe not but just a quick thought

**Comment by XD-DENG** [comment]
> maybe we don't need multiple watchers. maybe a single watcher can watch multiple namespaces.

I did think through of this idea earlier, but I couldn't think of a way to using existing K8S API to efficiently monitor multiple namespaces.

**Comment by dstandish** [comment]
> > maybe we don't need multiple watchers. maybe a single watcher can watch multiple namespaces.
> 
> I did think through of this idea earlier, but I couldn't think of a way to using existing K8S API to efficiently monitor multiple namespaces.

Maybe with threads? just spitballing here

**Comment by dstandish** [comment]
> > > maybe we don't need multiple watchers. maybe a single watcher can watch multiple namespaces.
> > 
> > 
> > I did think through of this idea earlier, but I couldn't think of a way to using existing K8S API to efficiently monitor multiple namespaces.
> 
> Maybe with threads? just spitballing here

or maybe we use multiple watchers, but run in thread instead of process

**Comment by potiuk** [comment]
Yep. Rebuild the image :)

`breeze ci-image build`

File: airflow/config_templates/config.yml [hunk 14]
```diff
@@ -2422,11 +2422,21 @@
...
       version_added: 1.10.12
       type: boolean
       example: ~
       default: "False"
+    - name: multi_namespace_mode_namespace_list
+      description: |
+        If multi_namespace_mode is True while scheduler does not have a cluster-role,
+        give the list of namespaces where the scheduler will schedule jobs
+        Scheduler needs to have the necessary permissions in these namespaces.
+      version_added: 2.5.1
```
Comment (dstandish) [inline airflow/config_templates/config.yml]: i think it will need to go in 2.6 since new feature

Comment (XD-DENG) [reply airflow/config_templates/config.yml]: Yep I agree. Will update to 2.6.0

(I just found this was missed to be updated so updating it to a "random" value and see if there is any luck it's relevant the weird CI errors. Obviously no luck)

Comment (XD-DENG) [reply airflow/config_templates/config.yml]: Updated to 2.6.0 at https://github.com/apache/airflow/pull/28047/commits/974e34521e6d32db7f50ca810c1f602e918f8eec

Continuing to check the CI failures.

**Comment by XD-DENG** [comment]
Hi @potiuk , actually along preparing this PR I have been using Breeze to run the tests locally, especially the `Core` as it's what this PR mainly (or only) touches. It has been running successfully and pass locally. I have tried to run it locally using Breeze again, and it still can succeed (` breeze testing tests --test-types Core --run-in-parallel --parallelism 32`)

Screenshot 2022-12-08 at 22 38 28

But it then always fail in the CI.

Any further input you can share to debug on this? Especially if any way to locate which certain test class/unit is causing the failure. Thanks a lot!

@dstandish @uranusjr  if you have any tips which may potentially help here, kindly let me know. Cheers!

**Comment by potiuk** [comment]
I think you should see at the memory usage while you are running it. For me it looks like some of the modifications there use accidentally vast amounts of memory @XD-DENG . That blows off our parallel running tests in ci (we have 64 GB of memory and tests are running in parallel,  so you do not have all 64GB available. 

And if this is correct assesment, then this is likely not only a CI but "Real" issue that might impact our users, so we better diagnose it. 

What I can do (and will now) is to assign "debug ci resources" label to this PR and rebase/re-run it. This one produces a nice output of resources that are being used while the test is running - so we will see what the memory usage is and be sure that this is the case (but 137 always so far indicated memory problems). 

BTW. In the future as commiter you can also do that @XD-DENG  :D - simply assigning the "debug ci resources" label before PR build is run will turn the resource monitoring on. 

Not sure what the reason for the alledged memory usage is but ;ater on, when this one gets fixed (hopefully?) you can also assign "use public runners"  labels and rebase/re-push the change and in this case it will run the same test on much more memory-constrained public runners that have far less memory (8GB) but tests there are run with far less parallelism so we can see if the change does not trigger a bit more "moderate" memory increase that might impact our non-commiter contributors.

**Comment by potiuk** [comment]
Added the "debug ci resources" label and re-run it. Let's see where it gets us. You will be able to see the regular dumps of memory int the task outputs while the tests are running (and if you run them locally you can also track it in similar way).

**Comment by potiuk** [comment]
See for example in the "wait for CI images" job:

image

**Comment by potiuk** [comment]
And even if you do not have as big of a machine you can control number of parallel tests by ``--parallelism`` flag to stress your system even more.

**Comment by potiuk** [comment]
More options available (such as ``-debug-resources` in `--help` output:

image

**Comment by potiuk** [comment]
Yep @XD-DENG - precisely as I expected:

At some point in time those tests start eating memory at the right of > 1GB / 10 seconds. Just before the crash we have just 800 MB left, Something spins out of control and causes that - this only happens on your change, not all the other PRs or main -  so it **MUST** be something here that triggers it. Now we just need to find out what it might be:

image

**Comment by potiuk** [comment]
My guess is - this is problem with cleanup. You have a LOT of parameterized tests and they are not cleaned up after each test.

**Comment by XD-DENG** [comment]
Thanks a lot @potiuk  for helping check and sharing the Breeze/CI tips!
I will do another check later today.

**Comment by potiuk** [comment]
Actually I think we need a bit more protection, otherwise the same situation happens if for any reason those asserts will start to raise exceptions. try/finally and making sure that we always .end() after we .start() acrosss all the k8s tests is a much more robust solution. 

Follow up here https://github.com/apache/airflow/pull/28281 @XD-DENG

**Comment by XD-DENG** [comment]
Hi @dstandish , for the File Task Handler issue we discussed earlier, I'm preparing the fix and should be ready within this week. FYI
